File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        queue<pair<int,int>> q;
        int m = image.size();

        int n = image[0].size();
        q.push({sr,sc});
        vector<vector<int>> visited(m,vector<int>(n,0));
        visited[sr][sc]=1;
        int originalColor = image[sr][sc];
        vector<pair<int,int>> displacement = {{-1,0},{1,0},{0,1},{0,-1}};
        while(!q.empty()) {
            auto top = q.front();
            int y = top.first;
            int x = top.second;
            q.pop();
            image[y][x]=color;
            for(auto disp:displacement) {
                int dy = disp.first;
                int dx = disp.second;

                if(x+dx>=0 && x+dx<n && y+dy>=0 && y+dy<m) {
                    if(image[y+dy][x+dx]==originalColor && visited[y+dy][x+dx]==0) {
                        visited[y+dy][x+dx]=1;
                        q.push({y+dy,x+dx});
                    }
                }
            }

        }
        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.py
from collections import deque
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        m,n = len(image),len(image[0])
        queue = deque()
        visitedSet = set()
        queue.append((sr,sc))

        originalColor = image[sr][sc]
        while queue:
            currR,currC = queue.popleft()
            image[currR][currC] = color
            for dR,dC in [(0,1),(0,-1),(1,0),(-1,0)]:
                newR,newC = dR+currR, dC+currC
                if newR==-1 or newC==-1 or newR==m or newC==n: continue
                if (newR,newC) in visitedSet or image[newR][newC]!=originalColor: continue
                visitedSet.add((newR,newC))
                queue.append((newR,newC))


        return image

                


            
        